doc/sphinx/practical-tools/utilities.rst

```diff
@@ -559,43 +629,73 @@ To build, say, two targets foo.vo and bar.vo in parallel one can use
 
 .. note::
 
-  ``make foo.vo bar.vo -j`` has a different meaning for the make
+  ``make foo.vo bar.vo -j`` has a different meaning for the ``make``
   utility, in particular it may build a shared prerequisite twice.
 
-
-.. note::
-
-  For users of coq_makefile with version < 8.7
-
-  + Support for "subdirectory" is deprecated. To perform actions before
-    or after the build (like invoking ``make`` on a subdirectory) one can hook
-    in pre-all and post-all extension points.
-  + ``-extra-phony`` and ``-extra`` are deprecated. To provide additional target
-    (``.PHONY`` or not) please use ``CoqMakefile.local``.
-
 .. note::
 
   Due to limitations with the compilation chain, makefiles generated
   by ``coq_makefile`` won't correctly compile OCaml plugins with OCaml
   < 4.07.0 when using more than one job (``-j N`` for ``N > 1``).
 
+
+The grammar of _CoqProject
+++++++++++++++++++++++++++
+The parser of ``coq_makefile`` uses the following syntax:
+
+  .. prodn::
+     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
+     blank ::= {| space | horizontal_tab | newline }
+     comment ::= # {* comment_char } newline
+     quoted_string ::= " {* quoted_char } "
+     unquoted_string ::= string_start_char {* unquoted_char }
+
+Where the following definitions apply:
+
+* :n:`space`, :n:`horizontal_tab` and :n:`newline` stand for the corresponding
+  ASCII characters.
+* :n:`comment_char` is the set of all characters except :n:`newline`.
+* :n:`quoted_char` is the set of all characters except ``"``.
+* :n:`string_start_char` is the set of all characters except those that match :n:`@blank`, or are ``"`` or ``#``.
+* :n:`unquoted_char` is the set of all characters except those that match :n:`@blank` or are ``#``.
```

Comment:
The grammar is not helpful.  For example, instead of

```
     comment ::= # {* comment_char } newline

* :n:`comment_char` is the set of all characters except :n:`newline`.
```

It's much more readable to say "Comments begin with a # and continue to the end of the line."

Please rework this part.

Reply:
Would it make more sense to move this grammar into coqProject_file.ml itself? Since I think such a grammar is a big help in understanding how [`parse`](https://github.com/Columbus240/coq/blob/bb7061659bc4dc074ba844bd231884586988112a/lib/coqProject_file.ml#L73-L124) works.

If the grammar were removed from the refman, I’d remove the rest of the paragraph as well, since that part can be inferred from the code.

Reply:
That would be reasonable.

Reply:
Still thinking about this and how to do it.

Reply:
I couldn't format the BNF satisfactorily in ocamldoc and didn't come up with an alternative description of the grammar that I find adequate. Either the text came out imprecise, overly wordy or introducing more concepts.

I think it's fine leaving it as it is. The intuitive description of the syntax is spread over the other parts of the doc. So an ordinary user probably won't gain anything from the grammar-paragraph. But for devs of other programs that parse `_CoqProject`, a formal description is all kinds of useful. It would make the file format look a little less ad-hoc and will help finding and fixing differences in behaviour (for example in PG or find-bugs.py).

Reply:
It could be ordinary comments rather than ocamldoc.  Unfortunately, it's very easy for formal descriptions to get out of date, so it's often necessary to verify the description is still current.  We can put this in for now, but it may be reworked when the section is eventually revised.